### PR TITLE
Remove `clearable` from select fields

### PR DESCRIPTION
**Pull request by andrearutrecht**
The `clearable` prop is removed from `FinalFormSelect` and `FinalFormAsyncSelect`. They are now clearable by default when `required` is not set.

---

After discussions with UX, the initial scope of this PR has changed. The planned placeholder item has been discarded.

## PR Checklist

-   [x] Verify if the change requires a changeset. See [CONTRIBUTING.md](https://github.com/vivid-planet/comet/blob/HEAD/CONTRIBUTING.md)
-   [x] Link to the respective task if one exists: SVK-172, SVK-173
-   [x] Provide screenshots/screencasts if the change contains visual changes

<details><summary>Screencasts</summary>
<p>
Behaviour of FinalFormSelect Story:

The FinalFormAsyncSelect story is not shown, as it is not functional at the moment. See COM-970

https://github.com/user-attachments/assets/28ca38d4-014b-472e-82b8-679a5b83536a

Behaviour of generated Form:

https://github.com/user-attachments/assets/63f1e861-7ea1-4d32-ab2f-92a545a8177c

</p>
</details>

### CODE REVIEW

File: demo/admin/src/products/future/generated/CreateCapProductForm.tsx [hunk 1]
```diff
@@ -121,7 +122,7 @@ export function CreateCapProductForm({ type }: FormProps): React.ReactElement {
                                         }
                                     `,
                                 });
-                                return data.productCategories.nodes;
+                                return [{ id: "", title: "Please select" }, ...data.productCategories.nodes];
```
Comment (johnnyomair) [inline demo/admin/src/products/future/generated/CreateCapProductForm.tsx:125]: title should be translatable

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts [hunk 2]
```diff
@@ -159,6 +161,13 @@ export function generateFormField(
             ${validateCode}
             {(props) =>
                 <FinalFormSelect ${config.readOnly ? readOnlyPropsWithLock : ""} {...props}>
+                ${
+                    !required
+                        ? `<MenuItem value="" >
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateFormField.ts:166]:
```suggestion
                        ? `<MenuItem value="">
```

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts [hunk 3]
```diff
@@ -159,6 +161,13 @@ export function generateFormField(
...
+                            <FormattedMessage id="${instanceGqlType}.${name}.pleaseSelect" defaultMessage="Please select" />
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateFormField.ts:167]: We should have a generic message for this in https://github.com/vivid-planet/comet/blob/next/packages/admin/admin/src/messages.ts

File: packages/admin/admin/src/form/FinalFormSelect.tsx [hunk 4]
```diff
@@ -5,12 +5,13 @@ import { FormattedMessage } from "react-intl";
 
 import { ClearInputAdornment } from "../common/ClearInputAdornment";
 import { AsyncOptionsProps } from "../hooks/useAsyncOptionsProps";
+import { messages } from "../messages";
 
 export interface FinalFormSelectProps<T> extends FieldRenderProps<T, HTMLInputElement | HTMLTextAreaElement> {
     getOptionLabel?: (option: T) => string;
     getOptionValue?: (option: T) => string;
     children?: React.ReactNode;
-    clearable?: boolean;
```
Comment (johnnyomair) [inline packages/admin/admin/src/form/FinalFormSelect.tsx:13]: This is a breaking change and requires a changeset

File: packages/admin/admin/src/form/FinalFormSelect.tsx [hunk 5]
```diff
@@ -28,23 +29,24 @@ export const FinalFormSelect = <T,>({
     },
     getOptionValue = (option: T) => {
         if (typeof option === "object" && option !== null) {
-            if ((option as any).id) return String((option as any).id);
+            // if ((option as any).id) return String((option as any).id);
```
Comment (johnnyomair) [inline packages/admin/admin/src/form/FinalFormSelect.tsx:32]:
```suggestion
```

There's no use in keeping this commented out code

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts [hunk 6]
```diff
@@ -150,7 +150,8 @@ export function generateFormField(
             ${required ? "required" : ""}
             fullWidth
             name="${name}"
-            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}>
+            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
+            >
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateFormField.ts:154]:
```suggestion
            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}>
```

File: storybook/src/admin/form/Select.tsx [hunk 7]
```diff
@@ -62,18 +61,6 @@ function Story() {
                                         )}
                                     </Field>
 
-                                    <Field name="flavorClearable" label="Clearable Flavor" fullWidth>
```
Comment (johnnyomair) [inline storybook/src/admin/form/Select.tsx:65]: Do we have a story to show an optional select field?

Comment (andrearutrecht) [reply storybook/src/admin/form/Select.tsx:65]: Any select field that is not required shows this behaviour. (Minus the multiple select field, see my [question](https://github.com/vivid-planet/comet/pull/2294#issuecomment-2245246215) above)

File: packages/admin/admin/src/form/FinalFormSelect.tsx (hunk 4 above)
Comment (andrearutrecht) [reply packages/admin/admin/src/form/FinalFormSelect.tsx:13]: Should there be two changesets? One for the removal of `clearable` and one for adding the placeholder item?

**Comment by johnnyomair** [comment]
> @johnnyomair I shortly discussed the default behaviour of optional multiple SelectFields with @jamesricky : We decided to not show the default item for these selects. They would be part of the selected value array. Is that okay with you?

Maybe show it while there is no other option selected? That would seem the most logical option to me.

**Comment by andrearutrecht** [comment]
> > @johnnyomair I shortly discussed the default behaviour of optional multiple SelectFields with @jamesricky : We decided to not show the default item for these selects. They would be part of the selected value array. Is that okay with you?
> 
> Maybe show it while there is no other option selected? That would seem the most logical option to me.

According to UX (see UX Meeting Notes of 2024-KW8), the "Please select" item should be a selectable item in the options dropdown. Should we deviate from that for multiple selects?

File: packages/admin/admin/src/form/FinalFormSelect.tsx (hunk 4 above)
Comment (johnnyomair) [reply packages/admin/admin/src/form/FinalFormSelect.tsx:13]: You could add a second line to the existing changeset regarding the new behavior with the empty option.

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts [hunk 8]
```diff
@@ -220,6 +220,7 @@ export function generateFormField(
                 fullWidth
                 name="${name}"
                 label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
+                ${!required ? "displayEmpty" : ""}
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateFormField.ts:223]: Is this even necessary any longer? Isn't this something the select adds by default?

File: packages/admin/cms-admin/src/generator/future/generateForm.ts [hunk 9]
```diff
@@ -204,6 +204,7 @@ export function generateForm(
         FinalFormSubmitEvent,
         Loading,
         MainContent,
+        messages,
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateForm.ts:207]:
```suggestion
```

This import isn't used anymore

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts (hunk 8 above)
Comment (andrearutrecht) [reply packages/admin/cms-admin/src/generator/future/generateFormField.ts:223]: No, it is required. The id of the placeholder item is and empty string, so it is not added to the form values. Without `displayEmpty`, the option is not displayed when selected.

**Comment by nsams** [comment]
Do we really need new "Please select" option PLUS the clear button?

**Comment by jamesricky** [comment]
> Do we really need new "Please select" option PLUS the clear button?

In a discussion with UX we decided selects should only have the "Please select" option to clear the value which would replace the clear button. So the clear button should be removed. 

> And have you (and UX) considered Selects used eg. in a filter form? (At least the "Please select" text needs to be changeable)

Not that I am aware of but that should be easy with something like an `clearOptionText` prop that has "Please select" as the default.

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts (hunk 8 above)
Comment (johnnyomair) [reply packages/admin/cms-admin/src/generator/future/generateFormField.ts:223]: I meant it's already added in the FinalFormSelect component: https://github.com/vivid-planet/comet/pull/2294/files/14ad7abd6685275016f681680290a1216769e837#diff-ef93179fe4c6d99e9e776fb1e928ac8b5af12c27fa2d2a6ea124e46028fb8b95R109. Shouldn't this be enough?

**Comment by johnnyomair** [comment]
> And have you (and UX) considered Selects used eg. in a filter form? (At least the "Please select" text needs to be changeable)

Do you mean the data grid filter? If not, could you give an example for a filter form?

**Comment by johnnyomair** [comment]
@andrearutrecht We discussed this today with UX: Every optional select should have a clear button, required selects shouldn't have a clear button (same behavior as for date pickers). The placeholder item should be removed. Could you please adapt this PR?

File: packages/admin/cms-admin/src/generator/future/generateFormField.ts (hunk 8 above)
Comment (andrearutrecht) [reply packages/admin/cms-admin/src/generator/future/generateFormField.ts:223]: Oh, you are right. Thanks for noticing!

**Comment by andrearutrecht** [comment]
> The placeholder item should be removed. 
Meaning the placeholder item should be completely removed from select fields?

**Comment by johnnyomair** [comment]
> Meaning the placeholder item should be completely removed from select fields?

Yes

File: .changeset/brown-cups-exist.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+---
+"@comet/cms-admin": major
```
Comment (johnnyomair) [inline .changeset/brown-cups-exist.md:2]:
```suggestion
```

This is now an `@comet/admin`-only change

File: packages/admin/admin/src/form/FinalFormSelect.tsx [hunk 11]
```diff
@@ -36,15 +36,15 @@ export const FinalFormSelect = <T,>({
         }
     },
     children,
-    clearable,
+    required,
```
Comment (jamesricky) [inline packages/admin/admin/src/form/FinalFormSelect.tsx:39]: Was `required` previously included in `...rest`? 
Then it needs to be added to `selectProps` so it's still passed to the `<Select />`.

Comment (andrearutrecht) [reply packages/admin/admin/src/form/FinalFormSelect.tsx:39]: Yes, it was missing. It's added now

**Review by jamesricky — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**